BLE GATT Object Transfer Service

kapi-no opened the pull request:
PR changes:
- Added Object Transfer Service library
- Added Peripheral OTS sample which demonstrates how to use service library.

kapi-no committed:
samples: bluetooth: adding peripheral_ots sample

Added a Bluetooth Peripheral sample that demonstrates how to use
GATT Object Transfer Service.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

kapi-no committed:
bluetooth: services: add Object Transfer service

Added GATT Object Transfer Service implementation.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

Thalley commented on subsys/bluetooth/services/ots_oacp.c:
```diff
@@ -0,0 +1,339 @@
+	}
+};
+
+static enum bt_gatt_ots_oacp_res_code oacp_command_decode(
+	const uint8_t *buf, uint16_t len, struct bt_gatt_ots_oacp_proc *proc)
+{
+	const uint8_t *end_buf = (buf + len);
+
+	proc->type = *buf++;
```
Consider using the `net_buf_simple` struct and functions to pull the data, rather than manually calling `sys_get_*` and `buf += ...`

joerchan replied:
I think `net_buf_simple_init_with_data` would be useful here.

Thalley commented on subsys/bluetooth/services/ots_oacp.c:
```diff
@@ -0,0 +1,339 @@
+	case BT_GATT_OTS_OACP_PROC_CREATE:
+		proc->create_params.size = sys_get_le32(buf);
+		buf += sizeof(proc->create_params.size);
+		if ((end_buf - buf) ==
+			sizeof(proc->create_params.type.uuid_16)) {
+			proc->create_params.type.uuid.type = BT_UUID_TYPE_16;
+			proc->create_params.type.uuid_16.val =
+				sys_get_le16(buf);
+		} else if ((end_buf - buf) ==
+			sizeof(proc->create_params.type.uuid_128)) {
+			proc->create_params.type.uuid.type = BT_UUID_TYPE_128;
+			memcpy(proc->create_params.type.uuid_128.val, buf,
+				sizeof(proc->create_params.type.uuid_128.val));
+		}
```
I wondered if it would be possible to simply do 
`memcpy(proc->create_params.type.uuid_16.val, buf, (end_buf - buf));` as the two are a union (after of course checking that the remaining length is either 16 or 128 bits.

joerchan replied:
I think `bt_uuid_create` would be useful here.

Thalley commented on subsys/bluetooth/services/ots_oacp.c:
```diff
@@ -0,0 +1,339 @@
+		LOG_DBG("OACP Read Op over L2CAP is completed");
+
+		if (read_op->sent_len > read_op->oacp_params.len) {
+			LOG_ERR("More bytes sent that the client requested");
```
```suggestion
			LOG_WARN("More bytes sent that the client requested");
```

Thalley commented on subsys/bluetooth/services/ots_olcp.c:
```diff
@@ -0,0 +1,313 @@
+
+	/* Prepare indication parameters */
+	memset(&ots->olcp_ind.params, 0, sizeof(ots->olcp_ind.params));
+	memcpy(&ots->olcp_ind.attr, olcp_attr, sizeof(ots->olcp_ind.attr));
```
This is just a pointer, isn't is not? I guess a simple assign would give the same result?

kapi-no replied:
No, it's a structure holding attribute content. I copy it here because I found no other way to pass OTS instance context to the indication callback.

Thalley replied:
Because `BT_GATT_CCC_MANAGED` doens't have support for user_data? 

But if `olcp_attr` is the write attribute, which does have it as you do `struct bt_gatt_ots *ots = (struct bt_gatt_ots *) oacp_attr->user_data;`, wouldn't `ots->oacp_ind.params.attr = olcp_attr;` give the same result? Or am I missing something here?

kapi-no replied:
yes, I tried doing it as you said but it did not work, unfortunately. Upon further investigation, it turned out that ``olcp_attr`` points to an automatic variable which becomes invalid (overwritten memory by other variables) in the indication callback context.

Thalley replied:
I see; that's rather surprising. But then yes, doing this copy would be a way of getting the OTS content. 

I've struggled with the same challenge, where my solution had been to iterate over all the attributes of all the service instances to find the pointer that matches the pointer in the indication callback, but this is likely more efficient just by storing some data.

joerchan replied:
Couldn't we use CONTAINER_OF in some way to get around this? I don't remember all the details but I think we solved this problem for HIDS using BT_GATT_CCC_INITIALIZER (although it goes through the gatt_pool.

kapi-no replied:
CONTAINER_OF is used in the indication callback. However, it works only if the attribute data is in static memory.

Thalley commented on subsys/bluetooth/services/ots_olcp_internal.h:
```diff
@@ -0,0 +1,95 @@
+#include <zephyr/types.h>
+#include <bluetooth/gatt.h>
+
+/**@brief The types of OLCP procedures. */
```
A general minor comment: I think it's customary to have spaces after `/**` and before `*/` in comments

Thalley reviewed (changes requested):
A couple of comments and change requests. Generally it seems to be covering a minimal viable implementation that can later be extended with the additional features of OTS.

joerchan commented on include/bluetooth/services/ots.h:
```diff
@@ -0,0 +1,600 @@
+#include <bluetooth/uuid.h>
+
+/** @brief Type of an OTS object. */
+struct bt_gatt_ots_obj_type {
```
Other services has already been added with the `bt_gatt` prefix, so I think we should continue to do so.

One thing I think we should resolve is how to distinguish between server and client versions. of the services.
At the moment we only have 3 services, HRS, DIS and BAS, all of them are server models.

In Bluetooth Mesh the models have the `_cli_` and `_srv_` infixes in the name, should we adopt this for the bluetooth services as well?

Thalley replied:
Good question, because there's a client version of OTS (dubbed OTC) coming, and if we want to start added _srv_ or _cli_ we might as well start now (unless we are just goint with ots/otc and bas/bac, etc.)

joerchan replied:
> unless we are just goint with ots/otc and bas/bac, etc

The `s` is for service though. I would suggest `_ots_srv_` and `_ots_cli_` similar to Bluetooth Mesh models.

Thalley replied:
That's true. Depending on the profile, the naming convention is sometimes e.g. [Feature] Server and [Feature] Client, although in this case, the the names given in the profile spec is just Object Transfer Service and Object Client. 

But I would be fine with the _srv_ and _cli_ naming convention.

kapi-no replied:
Would it then be ``bt_ots_srv_`` or ``bt_gatt_ots_srv_``? I prefer the first option because it is shorter.

Do you want me to align the naming in this PR?

Thalley replied:
@kapi-no I think it would probably be better to settle on a naming convention before you update all the code :) 

I agree with the argument you have with the shorter name, but on the other hand I also see a reason why to prefix it with `gatt`. I don't have any strong feelings between the two naming prefix (`bt_ots_srv_` and `bt_gatt_ots_srv_`); both works for me and I can't find anything in the documentation that strictly forbids either.

kapi-no replied:
> @kapi-no I think it would probably be better to settle on a naming convention before you update all the code :)
> 

Yes, that's why I am asking about both variants (so that we agree on the naming).


>Do you want me to align the naming in this PR?
>

I probably wasn't precise enough in this question. Do you want me to align the naming in this PR or should it rather be a part of another PR which aligns naming for all the service libraries in Zephyr (bas, dis, ...)?

Thalley replied:
Ah, misunderstood you then :) I think we should include whatever naming scheme we settle on in the PR, rather than updating it later.

joerchan commented on include/bluetooth/services/ots.h:
```diff
@@ -0,0 +1,600 @@
+
+/** @brief Initialize the OTS Server.
+ *
+ *  @param ots      OTS Server instance
```
OTS server seems correct, service is the S in OTS and should not be repeated I think.

Thalley commented on include/bluetooth/services/ots.h:
```diff
@@ -0,0 +1,600 @@
+			   uint64_t id,
+			   const struct bt_gatt_ots_obj_init *init);
+
+	/** Object deleted callback
+	 *
+	 *  This callback is called whenever an object is deleted.
+	 *
+	 *  @param ots  OTS Server instance
+	 *  @param conn The connection that deleted the object or NULL if
+	 *              this request came from the server
+	 *  @param id   Object ID
+	 */
+	void (*obj_deleted)(struct bt_gatt_ots *ots, struct bt_conn *conn,
+			    uint64_t id);
```
Should the application be able to refuse a deletion?

kapi-no replied:
It is possible to make an undeletable object by setting its properties correctly. If we added an API to change the object properties dynamically then it wouldn't be necessary to change this callback. Does it seem to you like a good solution?

Thalley replied:
Right, if the application wants to make an object undeletable, it can just set the properties of `bt_gatt_ots_obj_init` to that. Good point :) No need to update the API, unless there is a usecase for the application to perhaps only the device that created the object, delete it (e.g. by validating against the `conn`). 

Can imagine such a scenario where:
Client A creates ObjA, but where Client B is not allowed, by the application, to delete ObjA, but Client A is allowed to? 

If yes, then I think the API should allow for the application to deny a deletion, else it's good as it is.

kapi-no replied:
Let's keep it simple, OTS specification does not mention individual client permissions like superuser and standard one. I will tag this API as experimental so that we can change it in the future

kapi-no commented on include/bluetooth/services/ots.h:
```diff
@@ -0,0 +1,600 @@
+ */
+struct bt_gatt_attr *bt_gatt_ots_svc_get(struct bt_gatt_ots *ots);
+
+/** @brief Initialize the OTS Server.
```
S in OTS already stands for service, so I think we should skip this word.

kapi-no commented on subsys/bluetooth/services/Kconfig.ots:
```diff
@@ -0,0 +1,46 @@
+	int "Maximum number of available OTS instances"
+	default 1
+
+config BT_GATT_OTS_MAX_OBJ_CNT
+	int "Maximum number of objects that server can store"
+	default 5
```
It is per instance, perhaps I should add this in the description.

Defining a common object pool that is shared between instances could be added later on as another improvement. It could reduce the amount of memory required by this library in a multi-instance context when the number of objects, stored in each instance, is different.

kapi-no commented on subsys/bluetooth/services/ots.c:
```diff
@@ -0,0 +1,385 @@
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset,
+				 ots->cur_obj->metadata.name,
+				 strlen(ots->cur_obj->metadata.name) + 1);
```
Yes, it includes the null terminator, but I was not sure about this. It seems that the null terminator isn't mentioned in the OTS specification, so I can remove it in the next patch.

kapi-no commented on subsys/bluetooth/services/ots_internal.h:
```diff
@@ -0,0 +1,103 @@
+	uint32_t alloc;
+};
+
+struct bt_gatt_ots_obj_metadata {
+	uint64_t id;
+	char *name;
+	struct bt_gatt_ots_obj_type type;
+	struct bt_gatt_ots_obj_size size;
+	uint32_t props;
+};
```
I'm planning to get rid of it and use the one from the public header (also rename it as you suggested in the previous comment). Additionally, I will move ``id`` out of it since it is a special type of metadata that is frequently used to reference objects.

kapi-no commented on subsys/bluetooth/services/ots_internal.h:
```diff
@@ -0,0 +1,103 @@
+
+struct bt_gatt_ots_object_state {
+	enum bt_gatt_ots_object_state_type type;
+	union {
+		struct bt_gatt_ots_object_read_op {
+			struct bt_gatt_ots_oacp_read_params oacp_params;
+			uint32_t sent_len;
+		} read_op;
+	};
```
For now, we don't need it here. I just wanted to show my intention on how I would like this data structure to be extended in the future when more operations are supported by ots (e.g. write operation).

joerchan commented on subsys/bluetooth/services/ots_l2cap.c:
```diff
@@ -0,0 +1,208 @@
+/* According to BLE specification Assigned Numbers that are used in the
+ * Logical Link Control for protocol/service multiplexers.
+ */
+#define BLE_GATT_OTS_L2CAP_PSM	0x0025
```
BT_GATT

joerchan commented on subsys/bluetooth/services/ots_l2cap.c:
```diff
@@ -0,0 +1,208 @@
+#define BLE_GATT_OTS_L2CAP_PSM	0x0025
+
+/* Maximum size of TX buffer and its payload. */
+#define MAX_TX_BUF_SIZE		256
```
Could this perhaps be derived from the CONFIG_BT_GATT_OTS_L2CAP_CHAN_RX_MTU or L2CAP MPS?

kapi-no replied:
This is TX direction and RX_MTU can be different. I don't see any L2CAP MPS configuration.

I would rather not make it a configurable option since the OTS user doesn't need to know how data is actually sent over L2CAP. If he sends packets that is longer than 256 bytes (via ots_obj_read), then such a packet will be fragmented by this layer.

joerchan commented on subsys/bluetooth/services/ots_oacp.c:
```diff
@@ -0,0 +1,339 @@
+	struct bt_gatt_ots *ots = (struct bt_gatt_ots *) oacp_attr->user_data;
+
+	/* Encode OACP Response */
+	oacp_res[oacp_res_len++] = BT_GATT_OTS_OACP_PROC_RESP;
+	oacp_res[oacp_res_len++] = req_op_code;
+	oacp_res[oacp_res_len++] = oacp_status;
```
Maybe have a struct for this?

kapi-no replied:
I would rather not have one, it's one-time thing

joerchan commented on subsys/bluetooth/services/ots_oacp_internal.h:
```diff
@@ -0,0 +1,99 @@
+};
+
+/* Object Action Control Point procedure definition. */
+struct bt_gatt_ots_oacp_proc {
```
When using net_buf we usually don't need to have a this kind of struct with type and union of all possible packet types struct.
Instead we have just a simple struct, and pull from the buffer. You can have a look at hci command and events handling to what I mean (hci_core.c).

kapi-no replied:
I see, but it helps me to pass the data to dedicated functions inside OACP write callback and divide OACP processing into separate stages:
- ``oacp_command_decode`` (packet decoding)
- ``oacp_command_len_verify`` (length verification - ATT length error takes precedence before OTS errors)
- ``oacp_proc_validate`` (validation - checks if command is accepted by OTS)

In case of HCI Core, the HCI event type is identified, passed to the appropriate handler and processed there in a type-wise manner. Here, I use a stage-wise approach for all types:
- Initial validation
- Command decoding
- Command length verification
- OTS command validation (can it be executed?)
- Sending indication

joerchan reviewed (changes requested):
Sorry for the delay. Overall this looks very good!

Thalley commented on include/bluetooth/services/ots.h:
```diff
@@ -0,0 +1,600 @@
+	 *  Possible return values:
+	 *  -ENOMEM if no available space for new object.
+	 */
+	int (*obj_created)(struct bt_gatt_ots *ots, struct bt_conn *conn,
```
There is a control point for the client to create objects on the server. I guess the API is here to allow that, while also be used when the server creates an object, so that there's a common object create path.

kapi-no replied:
Yes, this is correct. @Thalley, thank you for explaining it.

Currently, this server implementation does not support operations for creating and deleting objects by the client, but I wanted to have uniform way of notifying application about created objects (regardless if the operation was triggered locally by the Server or remotely by the Client).

kapi-no commented on subsys/bluetooth/services/ots.c:
```diff
@@ -0,0 +1,385 @@
+	return 0;
+}
+
+SYS_INIT(bt_gatt_ots_instances_prepare, APPLICATION,
+	 CONFIG_APPLICATION_INIT_PRIORITY);
```
I`m planning to switch it to KERNEL_INIT_PRIORITY_DEFAULT

kapi-no commented on subsys/bluetooth/services/ots.c:
```diff
@@ -0,0 +1,385 @@
+		return BT_GATT_ERR(BT_GATT_OTS_OBJECT_NOT_SELECTED);
+	}
+
+	id.val = ots->cur_obj->metadata.id;
+	LOG_HEXDUMP_DBG(id.ba, sizeof(id.ba), "Current Object ID:");
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, id.ba,
+				 sizeof(id.ba));
```
I think you meant sys_put

Thalley replied:
No, I did mean `sys_get_le48`, as that would get the lower 48 bits of the uint64 ID and then return them, to be used as the value for bt_gatt_attr_read. In other words, it would look like

```
return bt_gatt_attr_read(conn, attr, buf, len, offset, sys_get_le48(id), 6);
```

kapi-no replied:
oh, ok. I would like to use sys_put for log_hexdump (getting rid of this ID struct).

kapi-no commented on include/bluetooth/services/ots.h:
```diff
@@ -0,0 +1,600 @@
+	 *  @return Data length to be sent via data parameter. This value
+	 *          should be smaller or equal to the len parameter.
+	 */
+	uint32_t (*obj_read)(struct bt_gatt_ots *ots, struct bt_conn *conn,
```
I don't think its needed here. The only erroneous case that I can think of would be when the application does not have the data that the client requests to read (should never happen).

Even then, the application could return 0 in the callback indicating that it does not have necessary data.

kapi-no commented on include/bluetooth/services/ots.h:
```diff
@@ -0,0 +1,600 @@
+ *
+ *  @return OTS instance in case of success or NULL in case of error.
+ */
+struct bt_gatt_ots *bt_gatt_ots_instance_get(void);
```
I`m planning to use _free_instance_get to make it compatible with docs.

kapi-no commented on subsys/bluetooth/services/ots.c:
```diff
@@ -0,0 +1,385 @@
+	return &BT_GATT_OTS_INSTANCE_LIST_START[instance_cnt++];
+}
+
+static int bt_gatt_ots_instances_prepare(struct device *dev)
```
Well, I wanted to distinguish it between init API function, which actually initializes OTS instance. This function simply ties together different memory chunks that constitute single OTS instance.

kapi-no commented on subsys/bluetooth/services/ots_l2cap.c:
```diff
@@ -0,0 +1,208 @@
+	return 0;
+}
+
+int bt_gatt_ots_l2cap_register(struct bt_gatt_ots_l2cap *l2cap_ctx)
```
Im planning to add it

kapi-no commented on subsys/bluetooth/services/ots.c:
```diff
@@ -0,0 +1,385 @@
+{
+	uint32_t oacp_ref = 0;
+
+	if (IS_ENABLED(CONFIG_BT_GATT_OTS_OACP_READ_SUPPORT)) {
```
Sure, this approach seems cleaner

kapi-no commented on samples/bluetooth/peripheral_ots/src/main.c:
```diff
@@ -0,0 +1,216 @@
+static int ots_obj_created(struct bt_gatt_ots *ots, struct bt_conn *conn,
+			   uint64_t id, const struct bt_gatt_ots_obj_init *init)
+{
+	uint32_t *id_trunc = (uint32_t *) &id;
```
Sure, I am planning to add inline function for converting binary Object ID to string (similar to BLE variant for addresses)

kapi-no commented on samples/bluetooth/peripheral_ots/src/main.c:
```diff
@@ -0,0 +1,216 @@
+#define DEVICE_NAME             CONFIG_BT_DEVICE_NAME
+#define DEVICE_NAME_LEN         (sizeof(DEVICE_NAME) - 1)
+
+#define BT_GATT_OTS_UNSPECIFIED_OBJ_TYPE 0x2ACA
```
Sure, I will move it there

kapi-no replied:
Decided to open an independent PR with it because I need to fix one existing OTS UUID declaration:

https://github.com/zephyrproject-rtos/zephyr/pull/27830

kapi-no committed:
bluetooth: uuid: adding object type identifiers for ots

Added Object Type identifiers for OTS.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

kapi-no commented on samples/bluetooth/peripheral_ots/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
```
Sure

kapi-no commented:
@Thalley, @joerchan, thank you for your reviews - I addressed all your comments.

I have been marking them as resolved so that it is easier for me to track remaining work items. However, if you think that something is still missing after my response or fix in code, you can unresolve your review comment.

kapi-no commented:
I tagged the OTS API as experimental so that we can still make changes to the API after it is merged.

kapi-no commented:
This PR contains commits from another PR in order to work:

https://github.com/zephyrproject-rtos/zephyr/pull/27830

They will be removed after linked PR is merged.

Thalley commented:
> This PR contains commits from another PR in order to work:
> 
> #27830
> 
> They will be removed after linked PR is merged.

One of them are also part of https://github.com/zephyrproject-rtos/zephyr/pull/27840

Thalley commented on subsys/bluetooth/services/ots/ots_srv_oacp.c:
```diff
@@ -0,0 +1,344 @@
+	if (err) {
+		LOG_ERR("L2CAP CoC error: %d while trying to execute OACP "
+			"Read procedure", err);
+		ots->cur_obj->state.type = BT_GATT_OTS_SRV_OBJECT_IDLE_STATE;
```
I wonder if the application would be able to delete the object in the `obj_read`, in which case I guess we need to check if `ots->cur_obj` is NULL here?

kapi-no replied:
It is not able to anymore (I return error when object is being read in the delete() API

Thalley commented on subsys/bluetooth/services/ots/ots_srv_oacp.c:
```diff
@@ -0,0 +1,344 @@
+	LOG_DBG("Executing Read procedure with offset: 0x%08X and "
+		"length: 0x%08X", params->offset, params->len);
+
+	if (ots->cb->obj_read) {
```
Didn't you add an __ASSERT for this? I guess the check can be removed, unless the client of course changes the callback during runtime...

kapi-no replied:
Well, I realized that supporting Object read operation is optional so I guess I will leave it here and correct my ASSERT logic.

Thalley commented on include/bluetooth/services/ots_srv.h:
```diff
@@ -0,0 +1,641 @@
+ *
+ *  @return 0 in case of success or negative value in case of error.
+ */
+int bt_gatt_ots_srv_object_delete(struct bt_gatt_ots_srv *ots, uint64_t id);
```
I wasn't properly able to determine whether this implementation supports that the application deletes an object while it is being read by a client, and what happens in that scenario. It's a corner case that maybe doesn't need to be handled in this version, but perhaps a TODO should be added if it isn't handled.

kapi-no replied:
Its easy to track now since each object has its own state so I added a guard statement that returns error when object is being used (non-idle state).

Thalley replied:
Alright, so the application is not able to delete objects that are being used. Sounds like a good and simple solution

Thalley reviewed (changes requested):
A few more comments

kapi-no commented on include/bluetooth/services/ots_srv.h:
```diff
@@ -0,0 +1,641 @@
+
+/** @brief OTS server callback structure. */
+struct bt_gatt_ots_srv_cb {
+	/** Object created callback
+	 *
+	 *  This callback is called whenever a new object is created.
+	 *  Application can reject this request by returning an error
+	 *  when it does not have necessary resources to hold this new
+	 *  object.
```
Yes, I`m planning to add this for both callbacks: obj_created and obj_deleted (this is also partially mentioned in the conn param description).

kapi-no commented on include/bluetooth/services/ots_srv.h:
```diff
@@ -0,0 +1,641 @@
+	 *  when it does not have necessary resources to hold this new
+	 *  object.
+	 *
+	 *  @param ots  OTS Server instance.
```
I am not sure about it, but you are right that the server is single entity.

Let's wait for feedback from @joerchan, because this will set a precedent on how we document remaining services that are going to be aligned with new naming convention (_srv).

How about this?
```suggestion
	 *  @param ots  Server-side OTS instance.
```

Thalley replied:
I would actually suggest to just have 
```
@param ots  OTS instance.
```

As that is, to me at least, very descriptive of what it is - A OT service instance

joerchan replied:
I'm not following. there can be many instances of `bt_gatt_ots_srv`, right? You can get one from the pool using `bt_gatt_ots_srv_free_instance_get`
I think it makes sense to include that it is an OTS server instance, since we could also have OTS client instances in the future.

Vudentz replied:
Isn't OTP the client side of here? Not sure why we would want to mix the terminologies of service with client, at least with regards to GATT a service already means a server so saying client instance here kind goes against that.

Thalley replied:
I guess it's a terminology discussion by now :)

My understanding is that there is 1 server, which may have one or more _service_ instances. 

Similarly on the client side: There's 1 client, but the client may interact with one or more _service_ instances.

kapi-no replied:
> 
> 
> I'm not following. there can be many instances of `bt_gatt_ots_srv`, right? You can get one from the pool using `bt_gatt_ots_srv_free_instance_get`

Yes, that's right.






> 
> 
> Isn't OTP the client side of here? Not sure why we would want to mix the terminologies of service with client, at least with regards to GATT a service already means a server so saying client instance here kind goes against that.

Yes, as according to OTP specification. However, we would like to develop client libraries using per service approach. For instance, HRP Client could be divided into dis_cli and hrs_cli instead of hrp_cli that would cover client-side operations for both services. This way, we can use these service client libraries in other profiles, because they are more modular. Also, the same service could be included in many profiles (e.g DIS in HID over GATT Profile and Heart Rate Profile). Apart from that, many simple services don't have profile specifications (e.g. DIS).

API naming was discussed in this conversation thread:
https://github.com/zephyrproject-rtos/zephyr/pull/27088#discussion_r459370693
and it was changed in this PR as we decided to follow the naming convention used in Mesh (health_cli & health_srv libraries). I hope that's okay for everyone.

Thalley replied:
It's a bit confusing since we have multiple threads discussing naming now :) 

If we want to use the same naming convention as Mesh, where it's named health_cli and health_src, perhaps it should be object (or simply obj) server and client, e.g. `obj_cli` and `obj_srv` (or `object_cli` and `object_srv`), which incidentally also matches the naming convention used in the OTP spec (see https://github.com/zephyrproject-rtos/zephyr/pull/27088#discussion_r481818288).

kapi-no replied:
It is "OTS instance" now, resolving

kapi-no commented on include/bluetooth/services/ots_srv.h:
```diff
@@ -0,0 +1,641 @@
+ *
+ *  @return Number of successfully formatted bytes from binary ID.
+ */
+static inline int bt_gatt_ots_srv_object_id_to_str(uint64_t object_id,
+						   char *str, size_t len)
+{
+	uint8_t id[6];
+
+	sys_put_le48(object_id, id);
+
+	return snprintk(str, len, "0x%02X%02X%02X%02X%02X%02X",
+			id[5], id[4], id[3], id[2], id[1], id[0]);
+}
```
I would leave it for now, it could be refactored in future PRs without breaking API since it's a new function.

BTW, it seems to me that ``bt_addr_le_str_real`` is not available as public API.

Thalley replied:
> BTW, it seems to me that bt_addr_le_str_real is not available as public API.
That is true (been an annoyance for me when I implement my test applications :P).  

This is indeed more similar to the public string functions such as `bt_uuid_to_str`.

kapi-no commented on include/bluetooth/services/ots_srv.h:
```diff
@@ -0,0 +1,641 @@
+	 *  @param ots    OTS Server instance.
+	 *  @param conn   The connection that read object.
+	 *  @param id     Object ID.
+	 *  @param data   In:  NULL once the read operations is completed.
```
Now, the application fully manages buffers that are used to send Object data. OTS notifies the upper layer when the read operation is finished so that the application can do whatever it wants with the buffer (e.g. update object content, free dynamic buffer, etc.)

kapi-no commented on subsys/bluetooth/services/ots/Kconfig.ots_srv:
```diff
@@ -0,0 +1,47 @@
+	select BT_GATT_DYNAMIC_DB
+	select BT_SMP
+	help
+	  Enable Object Transfer Service.
```
Well, I looked at the BT_EXT_ADV as an example experimental configuration and the experimental tag is used a bit earlier:
bool "(...) [EXPERIMENTAL]"
I am planning to add the tag as it is there.

Thanks for reminding me about it.

kapi-no commented on subsys/bluetooth/services/ots/Kconfig.ots_srv:
```diff
@@ -0,0 +1,47 @@
+if BT_GATT_OTS_SRV
+
+config BT_GATT_OTS_SRV_MAX_INST_CNT
+	int "Maximum number of available OTS instances"
+	default 1
```
Adding:
range 1 1 if !BT_GATT_OTS_SRV_SECONDARY_SVC
should help to enfore it.

kapi-no commented:
@Thalley, in the last patch, I addressed all comments except the unresolved one from your last review round.

Vudentz commented on subsys/bluetooth/services/ots/ots_srv.c:
```diff
@@ -0,0 +1,393 @@
+		_instance_num, BT_GATT_OTS_SRV_ATTRS_DEF, ots_instances)) \
+	static struct bt_gatt_service ots_service_list[] = {		  \
+		UTIL_LISTIFY(_instance_num, BT_GATT_OTS_SRV_SVC_ITEM)	  \
+	}
```
It would be great if we could generalize this to be able to define multiple instances in gatt.h (e.g. BT_GATT_SERVICE_INSTANCE), though it might be simpler to not have the objects such as bt_gatt_ots_srv and instead just pass the instance number to the callbacks so services can then resolve to its internal objects as they please.

Thalley replied:
I concur, that would be a very welcome addition

kapi-no replied:
Sure, I can add it, but I would rather do it in subsequent PRs. The scope of this one is already quite big.

This general macro part could accept another parameter _instances that contains whatever the user wants as a context in the service callbacks (whether it is an array of indexes or actual instances like bt_gatt_ots_srv). We can discuss it further when the PR is open.

joerchan replied:
Can it be added later without changing the API? Even though it is marked as experimental it is best to avoid, since it might propagate to downstream repositories that might use this API.

Vudentz replied:
It would be a new API/Macro, so we should not affect BT_GATT_SERVICE or relates APIs.

kapi-no replied:
There will be no API breaking either for OTS or in gatt.h. This instance definition macro is declared and used in the ots_srv.c file so the API user cannot access it.

joerchan commented on include/bluetooth/services/ots_srv.h:
```diff
@@ -0,0 +1,644 @@
+
+/** @brief OTS server callback structure. */
+struct bt_gatt_ots_srv_cb {
+	/** Object created callback
```
Missing `@brief` here and on the other callbacks.

joerchan commented on include/bluetooth/services/ots_srv.h:
```diff
@@ -0,0 +1,644 @@
+ *
+ *  @return 0 in case of success or negative value in case of error.
+ */
+int bt_gatt_ots_srv_object_add(struct bt_gatt_ots_srv *ots,
```
For consistency in the API i think it should be `bt_gatt_ots_srv_obj_add`, same for the other places where `_OBJECT_` or `_object_` is used.

kapi-no replied:
yes, it will be good to align this

joerchan commented on samples/bluetooth/peripheral_ots/prj.conf:
```diff
@@ -0,0 +1,7 @@
+CONFIG_BT_GATT_OTS_SRV=y
+
+CONFIG_LOG=y
+CONFIG_BT_GATT_OTS_SRV_LOG_LEVEL_DBG=y
```
Remove the log level set to DBG?

joerchan commented on include/bluetooth/services/ots_srv.h:
```diff
@@ -0,0 +1,644 @@
+
+	/* Callbacks */
+	struct bt_gatt_ots_srv_cb *cb;
+} __packed;
```
This shouldn't have packed here I think.

kapi-no replied:
yes, that's unnecessary

Vudentz commented on include/bluetooth/services/ots_srv.h:
```diff
@@ -0,0 +1,644 @@
+ */
+
+#ifndef ZEPHYR_INCLUDE_BLUETOOTH_SERVICES_OTS_SRV_H_
+#define ZEPHYR_INCLUDE_BLUETOOTH_SERVICES_OTS_SRV_H_
```
Btw, the S on OTS already means service so we could possible drop SRV term use on the APIs, etc.

joerchan replied:
SRV is for server, as opposed to OTS_CLI for client.

Vudentz replied:
Profiles are the ones that implement client side procedures, Services are always server side procedures, lets please not start mixing these up.

Thalley replied:
I guess the terminology is in fact that we have a Object Server that has one or more Object Transfer Services (defined by the OTS spec), and on the other side we have the Object Client that implements the procedures defined by the Object Transfer Profile. In reality service != server, and profile != client. 

We could rename things to be `OT_SRV` and `OT_CLI` or even `OBJ_SRV` and `OBJ_CLI` if we want to use the names by the OTP spec: 

![Screenshot_2020-09-02_09-12-38](https://user-images.githubusercontent.com/2520413/91943263-99889400-ecfc-11ea-91f9-8c28425ad840.png)

joerchan replied:
Thanks @Thalley. I agree that there does not seem to be a mapping from profile to client.
Also I would prefer to use the _cli_ and _srv_ to separate the client and server APIs to avoid confusion, and this pattern is already used in the Mesh models, which are similar to GATT services.

kapi-no replied:
This is resolved as part of a broader discussion on Slack

kapi-no commented:
I would appreciate if we could get this in before the Zephyr freeze. There are still two unresolved comments regarding naming discussion. I hope we can resolve them soon.

Vudentz commented on subsys/bluetooth/services/Kconfig.ots:
```diff
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config BT_GATT_OTS
+	bool "Object Transfer Service libraries"
```
Let mark this as experimental so we can fix this latter if needed.

kapi-no replied:
This is marked as experimental in the Kconfig that is located in OTS library. I am planning to merge both files into one and put the tag on the top-level configuration (that is BT_OTS)

kapi-no replied:
Now, the whole configuration is placed in one file services/ots/Kconfig with experimental tag attached to the BT_OTS definition

Vudentz commented on subsys/bluetooth/services/ots/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+  ots_srv_l2cap.c
+  ots_srv_obj_manager.c
+  ots_srv_oacp.c
+  ots_srv_olcp.c)
```
The directory is already ots, not sure why we need it prefixed to all files, though this can be fixed later.

kapi-no replied:
I'm planning to rename in the following way:

ots_srv.c -> server.c
ots_srv_l2cap.c -> l2cap.c
ots_srv_obj_manager.c -> obj_manager.c
ots_srv_oacp.c -> oacp_server.c
ots_srv_olcp.c  -> olcp_server.c

Plus headers:
ots_internal.h -> server_internal.h
ots_srv_l2cap_internal.h -> l2cap_internal.h
ots_srv_oacp_internal.h -> oacp_internal.h
ots_srv_olcp_internal.h -> olcp_internal.h
ots_srv_obj_manager_internal.h -> obj_manager_internal.h

kapi-no replied:
Well, I see some objections to my proposal so I am leaving it out from the scope of this PR as you suggested. This will be addressed in future PRs.

Thalley reviewed (approved):
I think all my comments have been addressed in a satisfactory manner, and the rest LGTM.

carlescufi reviewed (approved):
OK for the API and naming conventions.
